### PR TITLE
Add GUI module

### PR DESCRIPTION
**Pull request by Col-E**
## Preview

image

image

https://github.com/Minecraft-Malware-Prevention-Alliance/concoction/assets/21371686/d767fdf7-476f-425f-b2e4-ff3e3ad04910

Title says it all. Adding a common-use GUI.

## Pending contents

- [x] Progress reporting API in `lib` module _(Usage optional, but useful for UI feedback)_
- [x] Scanning UI implementation
    - [x] Core functionality
    - [x] Making it prettier
    - [x] Supporting exporting reports to file
    - [x] Supporting deleting files with detections
- [ ] Builder API for creating simplified UI's for specific malware campaigns _(pre-configured to target one thing, no user input really needed)_

### CODE REVIEW

File: concoction-gui/src/main/java/info/mmpa/concoction/panel/TableModelPanel.java [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+		inputTable = BasicComponentFactory.createTable(selectionModel, tableAdapter);
+		inputTable.setSelectionModel(new SingleListSelectionAdapter(selectionModel.getSelectionIndexHolder()));
+		JScrollPane tableWrapper = new JScrollPane();
+		tableWrapper.setViewportView(inputTable);
+
+		// Button actions for adding/removing content
+		FileDialog dialog = new FileDialog(context.getFrame(), bundle.getString("input.select-files"));
+		dialog.setMultipleMode(true);
+		dialog.setMode(FileDialog.LOAD);
+		dialog.setDirectory("C:\\Code\\Java\\concoction\\concoction-cli\\build\\libs");
```
Comment (Col-E) [inline concoction-gui/src/main/java/info/mmpa/concoction/panel/TableModelPanel.java:93]: I forgot to remove my testing code here 😆

**Comment by Col-E** [comment]
Huh, interesting failures. Works on my machine, and the failure on the builder test is really odd..

**Comment by Col-E** [comment]
GUI is basically done. Just need export support and management of actions on detected files.

Builder will then be done after that and re-use things from this GUI shown.

File: settings.gradle [hunk 2]
```diff
@@ -1,3 +1,4 @@
 rootProject.name = 'concoction'
 include('concoction-lib', 'concoction-cli')
-project(":concoction-lib").name = 'concoction'
\ No newline at end of file
+project(":concoction-lib").name = 'concoction'
+include 'concoction-gui'
```
Comment (xyzeva) [inline settings.gradle:4]: Why is this here? Move this to the `include` above.

File: concoction-lib/src/main/java/info/mmpa/concoction/scan/model/dynamic/DynamicMatchingModel.java [hunk 3]
```diff
@@ -37,41 +37,33 @@ public DynamicMatchingModel(@Nonnull Map<String, DynamicMatchEntry> variants) {
...
+				Unchecked.runSafe("dynamic-sink-feed", () -> sink.onDetection(path, archetype, new Detection(archetype, path)));
 	}
 
 	/**
 	 * @param sink
 	 * 		Sink to feed match results into.
 	 * @param archetype
 	 * 		Information about what the signature being matched.
-	 * @param sourcePath
+	 * @param path
```
Comment (xyzeva) [inline concoction-lib/src/main/java/info/mmpa/concoction/scan/model/dynamic/DynamicMatchingModel.java:57]: Do we want to change this? Seems less descriptive to me.

File: concoction-lib/src/main/java/info/mmpa/concoction/scan/model/dynamic/DynamicMatchingModel.java [hunk 4]
```diff
@@ -37,41 +37,33 @@ public DynamicMatchingModel(@Nonnull Map<String, DynamicMatchEntry> variants) {
...
 	 * 		Current method path to the containing input source. SSVM holds the rest of the details.
 	 * @param frame
 	 * 		SSVM frame of method exited.
 	 */
 	public void matchOnExit(@Nonnull ResultsSink sink, @Nonnull DetectionArchetype archetype,
-							@Nonnull SourcePathElement sourcePath, @Nonnull CallStackFrame frame) {
+							@Nonnull MethodPathElement path, @Nonnull CallStackFrame frame) {
```
Comment (xyzeva) [inline concoction-lib/src/main/java/info/mmpa/concoction/scan/model/dynamic/DynamicMatchingModel.java:63]: Do we want to change this? Seems less descriptive to me.

**Review by xyzeva — CHANGES_REQUESTED** [review]

File: concoction-lib/src/main/java/info/mmpa/concoction/scan/model/dynamic/DynamicMatchingModel.java (hunk 4 above)
Comment (Col-E) [reply concoction-lib/src/main/java/info/mmpa/concoction/scan/model/dynamic/DynamicMatchingModel.java:63]: The type is changed. This should only ever consume a method path.

File: concoction-lib/src/main/java/info/mmpa/concoction/scan/model/dynamic/DynamicMatchingModel.java (hunk 3 above)
Comment (Col-E) [reply concoction-lib/src/main/java/info/mmpa/concoction/scan/model/dynamic/DynamicMatchingModel.java:57]: Same as below comment

File: settings.gradle (hunk 2 above)
Comment (Col-E) [reply settings.gradle:4]: IntelliJ auto-add moment

File: concoction-lib/src/main/java/info/mmpa/concoction/scan/model/dynamic/DynamicMatchingModel.java (hunk 4 above)
Comment (xyzeva) [reply concoction-lib/src/main/java/info/mmpa/concoction/scan/model/dynamic/DynamicMatchingModel.java:63]: Could we change it to `methodPath` then?

File: concoction-lib/src/main/java/info/mmpa/concoction/Concoction.java [hunk 5]
```diff
@@ -63,6 +68,39 @@ public Concoction withMaxInputDirectoryDepth(int inputDepth) {
 		return this;
 	}
 
+	/**
```
[the remark this answers is not in the document]
Comment (Col-E) [reply concoction-lib/src/main/java/info/mmpa/concoction/Concoction.java:71]: The changes were tied to things required by this GUI work, and there's some overlap so cherry-picking would be a pain since it'd require editing portions of commits

File: concoction-lib/src/main/java/info/mmpa/concoction/scan/model/dynamic/DynamicMatchingModel.java (hunk 4 above)
Comment (Col-E) [reply concoction-lib/src/main/java/info/mmpa/concoction/scan/model/dynamic/DynamicMatchingModel.java:63]: Done

**Comment by Col-E** [comment]
Ok, Java 8 back-porting of Ikonli done. The GUI is not usable across the board when build with `:concoction-gui:shadowJar`

File: app-patching.gradle [hunk 6]
```diff
@@ -1,21 +1,95 @@
...
-        classpath 'com.github.RaphiMC.JavaDowngrader:core:13e29798a8'
+        // Some funky problems with gradle classpath management prevent us from compiling against the normal ASM dependency.
+        // So we use their old '-all' shaded version since that is technically a different artifact to get access
+        // to the ASM classes we need for patching.
+        // 
+        // When gradle is used at runtime, we use the appropriate dependency as specified in 'settings.gradle' which is
+        // an up-to-date version of ASM.
+        //
+        // I know, gradle is dumb.
+        classpath 'org.ow2.asm:asm-all:5.2'
```
Comment (xyzeva) [inline app-patching.gradle:15]: Move this to `dependencies.gradle`, electric boogaloo 😢

File: settings.gradle [hunk 7]
```diff
@@ -1,3 +1,15 @@
+buildscript {
+    repositories {
+        mavenCentral()
+    }
+
+    // Used for patching in the GUI module to use a Java 11 dependency on Java 8.
+    // See 'app-patching.gradle' for more details.
+    dependencies {
+        classpath 'org.ow2.asm:asm:9.5'
```
Comment (xyzeva) [inline settings.gradle:9]: Move to `dependencies.gradle`, tokyo drift 😿

**Review by xyzeva — CHANGES_REQUESTED** [review]

**Review by xyzeva — APPROVED** [review]
LGTM
